### PR TITLE
OpenBSD compatibility

### PR DESCRIPTION
**Pull request by nanolith**
Patches to support CBMC on OpenBSD.

The ansi-c library has been patched to provide a compatible pthread_barrier_init function, and to bypass OpenBSD's stdio.h, which can't currently be parsed by the checker script.

The c_preprocess.cpp source file in ansi-c has been patched to pass the appropriate -std to Clang on OpenBSD.  OpenBSD does not use GNU extensions with the system Clang.

The config.cpp source file in util has been patched so that OpenBSD defaults to c99, just like FreeBSD.

Patches have been protected with a `defined(__OpenBSD__)` check to ensure that they do not impact other builds.

### CODE REVIEW

File: src/ansi-c/library/stdio.c [hunk 1]
```diff
@@ -2,7 +2,13 @@
 /* FUNCTION: putchar */
 
 #ifndef __CPROVER_STDIO_H_INCLUDED
+# if !defined(__OpenBSD__)
 #include <stdio.h>
+# else
+// on OpenBSD, there are some definitions that the checker does not like.
+#include <stdio_openbsd_compat.h>
```
Comment (tautschnig) [inline src/ansi-c/library/stdio.c]: Would you mind sharing any details what's going wrong on OpenBSD with the stock header files?

**Review by tautschnig — APPROVED** [review]
Good to see portability to another platform! In a couple of places `clang-format` will be unhappy with the formatting, though. Please take a look at the logs of the failing CI tasks.

File: src/util/config.cpp [hunk 2]
```diff
@@ -677,7 +677,7 @@ configt::ansi_ct::c_standardt configt::ansi_ct::default_c_standard()
   #if defined(__APPLE__)
   // By default, clang on the Mac builds C code in GNU C11
   return c_standardt::C11;
-  #elif defined(__FreeBSD__)
+  #elif defined(__FreeBSD__) || defined(__OpenBSD__)
   // By default, clang on FreeBSD builds C code in GNU C99
   return c_standardt::C99;
```
Comment (kroening) [inline src/util/config.cpp:683]: Please tweak the comment to say "clang on FreeBSD and OpenBSD".

Comment (kroening) [reply src/util/config.cpp:683]: Also, the changes to c_preprocess.cpp suggest that the default isn't GNU C99 but actually proper standard C99?

File: src/ansi-c/library/stdio.c (hunk 1 above)
Comment (nanolith) [reply src/ansi-c/library/stdio.c]: Basically, putc/getc are macros in OpenBSD.  The expansion used in this source file can't be parsed by the checker.  Here is an example:

> Checking /home/jhandville/packages/cbmc/src/ansi-c/library/stdio.c
> __libcheck.c:11:13: error: expected identifier or '('
> inline int (!__isthreaded ? __sputc(int c, (&__sF[1])) : (putc)(int c, (&__sF[1])))
>             ^
> __libcheck.c:11:13: error: expected ')'
> __libcheck.c:11:12: note: to match this '('
> inline int (!__isthreaded ? __sputc(int c, (&__sF[1])) : (putc)(int c, (&__sF[1])))
>            ^
> __libcheck.c:312:13: error: expected identifier or '('
> inline int (!__isthreaded ? (((FILE *stream)->_flags & 0x0020) != 0) : (feof)(FILE *stream))
>             ^
> __libcheck.c:312:13: error: expected ')'
> __libcheck.c:312:12: note: to match this '('
> inline int (!__isthreaded ? (((FILE *stream)->_flags & 0x0020) != 0) : (feof)(FILE *stream))
>            ^

File: src/util/config.cpp (hunk 2 above)
Comment (nanolith) [reply src/util/config.cpp:683]: Yes.  OpenBSD does not use the GNU extensions AFAIK.  Either way, -std=gnuxx aren't valid options on OpenBSD.

File: src/ansi-c/library/stdio.c (hunk 1 above)
Comment (tautschnig) [reply src/ansi-c/library/stdio.c]: Would adding `#undef putc` and `#undef getc` do the trick? We already do this for some of the functions.

Comment (nanolith) [reply src/ansi-c/library/stdio.c]: That's a possibility.  Let me try.

File: src/util/config.cpp (hunk 2 above)
Comment (nanolith) [reply src/util/config.cpp:683]: Comment updated as requested.

File: src/ansi-c/library/stdio.c (hunk 1 above)
Comment (nanolith) [reply src/ansi-c/library/stdio.c]: Simplified with some #undefs as per your suggestion.

**Review by nanolith — COMMENTED** [review]
Added commits to address feedback.

**Comment by nanolith** [comment]
Is there anything else that the collaborators need from me for this PR?  I just want to confirm that you didn't need anything else from my end.

**Comment by tautschnig** [comment]
@nanolith Yes, please help `clang-format` be happy as https://travis-ci.org/diffblue/cbmc/jobs/488215218 shows. It's just about whitespace...

**Comment by nanolith** [comment]
Fixed all clang-format issues.  All green.  :-)

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3236da79f1571d67281e7a2cf98dce23410abf15).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99968034

**Pull request merged**